sdk/nodejs/cmd/pulumi-language-nodejs/proxy.go

```diff
@@ -0,0 +1,253 @@
+// runtime. One file is used by it to send us messages, and one file is used by us to send
+// messages back.  Because these are just files, nodejs natively supports allowing both sides to
+// read and write from each synchronously.
+//
+// When we receive the sync-invoke messages from the nodejs sdk we deserialize things off of the
+// file and then make a synchronous call to the real engine `invoke` monitor endpoint.  Unlike
+// nodejs, we have no problem calling this synchronously, and can block until we get the
+// response which we can then synchronously send to nodejs.
+func newMonitorProxy(
+	responseChannel chan<- *pulumirpc.RunResponse, targetAddr string,
+	tracingSpan opentracing.Span) (*monitorProxy, error) {
+
+	pipes, err := createPipes()
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := grpc.Dial(targetAddr, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	proxy := &monitorProxy{
+		target:        pulumirpc.NewResourceMonitorClient(conn),
+		pipeDirectory: pipes,
+	}
+
+	// Channel to control the server lifetime.  The pipe reading code will ask the server to
+	// shutdown once it has finished reading/writing from the pipes for any reason.
+	serverCancel := make(chan bool)
+	port, serverErrors, err := rpcutil.Serve(0, serverCancel, []func(*grpc.Server) error{
+		func(srv *grpc.Server) error {
+			pulumirpc.RegisterResourceMonitorServer(srv, proxy)
+			return nil
+		},
+	}, tracingSpan)
+	if err != nil {
+		return nil, err
+	}
+
+	proxy.addr = fmt.Sprintf("127.0.0.1:%d", port)
+
+	// Lister for errors from the server and push them to the singular error stream if we receive one.
+	go func() {
+		err := <-serverErrors
+		if err != nil {
+			responseChannel <- &pulumirpc.RunResponse{Error: err.Error()}
+		}
+	}()
+
+	// Now, kick off a goroutine to actually read and write from the pipes.  Any errors should be
+	// reported to `responseChannel`.  When complete, `serverCancel` should be called to let the
+	// server know it can shutdown gracefully.
+	go proxy.servePipes(responseChannel, serverCancel)
+
+	return proxy, nil
+}
+
+func createPipes() (string, error) {
+	dir, err := ioutil.TempDir("", "pulumi-node-pipes")
+
+	if err != nil {
+		return "", err
+	}
+	if err := syscall.Mkfifo(path.Join(dir, "invoke_req"), 0600); err != nil {
+		return "", err
+	}
+	if err := syscall.Mkfifo(path.Join(dir, "invoke_res"), 0600); err != nil {
+		return "", err
+	}
+
+	return dir, nil
+}
+
+func (p *monitorProxy) servePipes(
+	resultChannel chan<- *pulumirpc.RunResponse, serverCancel chan<- bool) {
+
+	// Once we're done using the pipes, let the server know it can shutdown gracefully.
+	defer func() {
+		serverCancel <- true
+	}()
+
+	// Keep reading and writing from the pipes until we run into an error or are canceled.
+	err := func() error {
+		pbcodec := encoding.GetCodec(proto.Name)
+
+		defer contract.IgnoreError(os.Remove(p.pipeDirectory))
```

Comment:
added cleanup logic to make sure we don't continue leaving cruft in the temp dir.